# Chapter 3: Data exploration

In [58]:
%reset
low_memory=False
import numpy as np
import pandas as pd

Once deleted, variables cannot be recovered. Proceed (y/[n])?  y


## 3.1 Introduction & Problem Setting

Today, you will tackle your very first data cleaning task! You will continue working on the Titanic dataset 🚢 and explore the many approaches you can use when cleaning a dataset. You will discover how each approach has its own advantages and disadvantages. Think of data cleaning like tidying up a messy room – different methods can help, but some are more effective than others!

## 3.2 Explore the content of the dataset

Load the Titanic dataset and start from the unmodified data. Explore the dataset and identify the following:

- How many records and columns do you have?
- How many columns contain missing values?
- How many columns do you want to transform?
- Which columns are these?

In [3]:
titanic = pd.read_csv("titanic_uncleaned.csv", sep = ";")
titanic.shape

(892, 12)

The .shape attribute returns the dimensions of the DataFrame as (rows, columns), giving us the total number of records and columns.

The .shape attribute returns the dimensions of the DataFrame as (rows, columns), giving us the total number of records and columns.

In [4]:
titanic.describe()

,PassengerId,Pclass,Age,SibSp,Parch
count,892.000000,891.000000,714.000000,891.000000,891.000000
mean,446.500000,2.308642,29.706120,0.523008,0.381594
std,257.642517,0.836071,14.523986,1.102743,0.806057
min,1.000000,1.000000,0.420000,0.000000,0.000000
25%,223.750000,2.000000,20.125000,0.000000,0.000000
50%,446.500000,3.000000,28.000000,0.000000,0.000000
75%,669.250000,3.000000,38.000000,1.000000,0.000000
max,892.000000,3.000000,80.000000,8.000000,6.000000


The .describe() method provides summary statistics for numeric columns, including count, mean, standard deviation, min, max, and quartiles.

In [5]:
titanic.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
count,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,2,2,2,1,2,2,2,2,2,2,2
top,False,False,False,False,False,False,False,False,False,False,True,False
freq,892,891,891,891,892,714,891,891,891,891,688,890


Applying .describe() to the boolean DataFrame from .isna() shows counts of missing (True) and present (False) values per column.

In [6]:
titanic.dtypes

PassengerId      int64
Survived        object
Pclass         float64
Name            object
Sex             object
Age            float64
SibSp          float64
Parch          float64
Ticket          object
Fare            object
Cabin           object
Embarked        object
dtype: object

The .dtypes attribute shows the data type of each column, helping identify which need conversion.

The dataset has 892 records and 12 columns.

Of those 12 columns, only two have no missing values: 'PassengerId' and 'Sex'.

All other columns have some missing values, with 'Age' and 'Cabin' having significantly more.

About half of the columns are already numeric. The ones we likely want to transform are:

- Survived

- Name

- Sex

- Age

- Ticket

- Cabin

- Embarked

## 3.3 Data transformations

Transform the data types of your columns! Make sure your data is in the desired form so you can easily handle the missing values later.

*Hint: We have not yet seen what to do with data in columns like 'Name', 'Ticket', and 'Cabin'. Thus, you can leave those three columns as they are.*

Transform the data types of your columns! Make sure your data is in the desired from so you can easily handle the missing values later.

*Hint: we have not yet seen what to do with data such as in the columns 'Name', 'Ticket' and 'Cabin'. Thus, you can leave those three columns as they are.*

In [7]:
titanic['Survived'] = titanic['Survived'].map({'No': 0, 'Yes' : 1})
titanic['Sex'] = titanic['Sex'].map({'male': 0, 'female' : 1})
titanic = pd.get_dummies(titanic, columns = ['Embarked'])
titanic['Embarked_C'] = titanic['Embarked_C'].astype('float')
titanic['Embarked_Q'] = titanic['Embarked_Q'].astype('float')
titanic['Embarked_S'] = titanic['Embarked_S'].astype('float')

The .map() method replaces values based on a dictionary. Here, it converts 'No'/'Yes' to 0/1 for 'Survived' and 'male'/'female' to 0/1 for 'Sex'.

pd.get_dummies() creates binary columns for each unique value in 'Embarked'.

.astype('float') converts the new dummy columns to float type for consistency.

In [8]:
titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
0,1,0.0,3.0,"Braund, Mr. Owen Harris",0,NaN,1.0,0.0,A/5 21171,7.25,NaN,0.0,0.0,1.0
1,2,1.0,1.0,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1.0,0.0,PC 17599,712.833,C85,1.0,0.0,0.0
2,3,1.0,3.0,"Heikkinen, Miss. Laina",1,26.0,0.0,0.0,STON/O2. 3101282,7.925,NaN,0.0,0.0,1.0
3,4,1.0,1.0,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1.0,0.0,113803,53.1,C123,0.0,0.0,1.0
4,5,0.0,3.0,"Allen, Mr. William Henry",0,35.0,0.0,0.0,373450,8.05,NaN,0.0,0.0,1.0


The .head() method displays the first few rows of the DataFrame to preview the transformed data.

## 3.4 Missing values

Now it's time to tackle the missing values. First, check what you may have already done. Compare the current data with the raw dataset. Did you already solve some missing values?

Now it's time to tackle the missing values. Fist of all, have a look at what you may have already done on this subject. Compare the data you currently have with the raw dataset. Did you already solve some missing values?

In [9]:
titanic.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,2,2,2,1,2,2,2,2,2,2,1,1,1
top,False,False,False,False,False,False,False,False,False,False,True,False,False,False
freq,892,891,891,891,892,714,891,891,891,891,688,892,892,892


This checks for missing values in the current dataset after transformations.

In [10]:
titanic[np.logical_and(np.logical_and(titanic.Embarked_C == 0, titanic.Embarked_Q == 0), titanic.Embarked_S == 0)]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
61,62,1.0,1.0,"Icard, Miss. Amelie",1,38.0,0.0,0.0,113572,80,B28,0.0,0.0,0.0
829,830,1.0,1.0,"Stone, Mrs. George Nelson (Martha Evelyn)",1,62.0,0.0,0.0,113572,80,B28,0.0,0.0,0.0


This filters rows where all Embarked dummy columns are 0, indicating original missing values that were handled by get_dummies.

By using the .get_dummies() function, we have already handled some missing values. The three new columns no longer have missing values; instead, they contain only 0s or 1s. Records with original missing 'Embarked' values now have 0 in all three columns.

In [31]:
titanic['Fare'].dtype

dtype('O')

Checking the data type of 'Fare' before cleaning.

In [43]:
def clean_fare(value):
    if len(value) > 3:
        if value[-3] == '.' and value[-2:].isdigit():
            value = value.replace('.', ',', 1)
        value = value.replace('.', '')
        value = value.replace(',', '.')
    return value

titanic['Fare'] = titanic['Fare'].apply(clean_fare)

This custom function cleans the 'Fare' values by handling European number formatting (commas as decimals) and converting to standard float format.

Now it's time to get to work! Take a copy of your cleaned dataset and handle the missing values according to the three ways you've seen. Analyze and describe the effect each method has on the data. How many rows do you retain, how many columns? How good is the data quality? Can you think of ways to improve your approach?

### 3.4.1: Data removal

In [11]:
titanic2 = titanic.dropna()

dropna() removes all rows with any missing values, which can significantly reduce the dataset.

In [12]:
titanic2.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,185,185,185,185,185,185,185,185,185,185,185,185,185,185
unique,1,1,1,1,1,1,1,1,1,1,1,1,1,1
top,False,False,False,False,False,False,False,False,False,False,False,False,False,False
freq,185,185,185,185,185,185,185,185,185,185,185,185,185,185


Simply dropping all records with missing values leads to a huge data loss.

Since the 'Cabin' column has mostly missing values, we can drop it entirely to preserve more records.

In [13]:
titanic2 = titanic.drop(columns=['Cabin'])
titanic2 = titanic2.dropna()

drop(columns=['Cabin']) removes the 'Cabin' column from the DataFrame.

In [14]:
titanic2.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked_C,Embarked_Q,Embarked_S
count,713,713,713,713,713,713,713,713,713,713,713,713,713
unique,1,1,1,1,1,1,1,1,1,1,1,1,1
top,False,False,False,False,False,False,False,False,False,False,False,False,False
freq,713,713,713,713,713,713,713,713,713,713,713,713,713


Now we retain 713 records, which is much better!

### 3.4.2: Data Imputation

In [15]:
titanic3 = titanic

Creating a copy of the DataFrame for imputation experiments.

In [16]:
titanic3.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,2,2,2,1,2,2,2,2,2,2,1,1,1
top,False,False,False,False,False,False,False,False,False,False,True,False,False,False
freq,892,891,891,891,892,714,891,891,891,891,688,892,892,892


ffill() fills missing values with the last known value from previous rows.

In [17]:
titanic3['Survived'] = titanic3['Survived'].ffill()
titanic3['Pclass'] = titanic3['Pclass'].ffill()
titanic3['Name'] = titanic3['Name'].ffill()
titanic3['Sex'] = titanic3['Sex'].ffill()
titanic3['Age'] = titanic3['Age'].ffill()
titanic3['SibSp'] = titanic3['SibSp'].ffill()
titanic3['Parch'] = titanic3['Parch'].ffill()
titanic3['Ticket'] = titanic3['Ticket'].ffill()
titanic3['Fare'] = titanic3['Fare'].ffill()
titanic3['Cabin'] = titanic3['Cabin'].ffill()

ffill() fills missing values with the last known value from previous rows.

In [18]:
titanic3.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,1,1,1,1,2,1,1,1,1,2,1,1,1
top,False,False,False,False,False,False,False,False,False,False,False,False,False,False
freq,892,892,892,892,892,891,892,892,892,892,891,892,892,892


We can see that the 'Age' and 'Cabin' columns still have missing values. Let's investigate!

In [25]:
titanic3.loc[np.logical_or(titanic3['Cabin'].isna(), titanic3['Age'].isna())]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
0,1,0.0,3.0,"Braund, Mr. Owen Harris",0,NaN,1.0,0.0,A/5 21171,7.25,NaN,0.0,0.0,1.0


This locates rows where 'Age' or 'Cabin' are still missing after ffill.

Aha, this makes sense! The first record has missing values, so there's no previous value to forward fill. We'll use bfill() to fill from subsequent values.

In [26]:
titanic3['Age'] = titanic3['Age'].bfill()
titanic3['Cabin'] = titanic3['Cabin'].bfill()
titanic3.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,1,1,1,1,1,1,1,1,1,1,1,1,1
top,False,False,False,False,False,False,False,False,False,False,False,False,False,False
freq,892,892,892,892,892,892,892,892,892,892,892,892,892,892


bfill() fills missing values with the next known value from subsequent rows.

### 3.4.3: Data Flagging

Unfortunately, Python creates copies by reference, so modifying the 'copy' affects the original. We need to reload the data.

In [59]:
titanic = pd.read_csv("titanic_uncleaned.csv", sep = ";")

titanic['Survived'] = titanic['Survived'].map({'No': 0, 'Yes' : 1})
titanic['Sex'] = titanic['Sex'].map({'male': 0, 'female' : 1})
titanic = pd.get_dummies(titanic, columns = ['Embarked'])
titanic['Embarked_C'] = titanic['Embarked_C'].astype('float')
titanic['Embarked_Q'] = titanic['Embarked_Q'].astype('float')
titanic['Embarked_S'] = titanic['Embarked_S'].astype('float')

def clean_fare(value):
    value = str(value)
    if len(value) > 3:
        if value[-3] == '.' and value[-2:].isdigit():
            value = value.replace('.', ',', 1)
        value = value.replace('.', '')
        value = value.replace(',', '.')
    return value

titanic['Fare'] = titanic['Fare'].apply(clean_fare).astype('float')

Reloading and reapplying transformations, then using fillna() to flag missing values with special numbers (-1 for numeric, "" for strings).

In [61]:
titanic['Survived'] = titanic['Survived'].fillna(-1)
titanic['Pclass'] = titanic['Pclass'].fillna(-1)
titanic['Name'] = titanic['Name'].fillna("")
titanic['Age'] = titanic['Age'].fillna(-1)
titanic['SibSp'] = titanic['SibSp'].fillna(-1)
titanic['Parch'] = titanic['Parch'].fillna(-1)
titanic['Ticket'] = titanic['Ticket'].fillna("")
titanic['Fare'] = titanic['Fare'].fillna(-1)
titanic['Cabin'] = titanic['Cabin'].fillna("")

In [62]:
titanic.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,1,1,1,1,1,1,1,1,1,1,1,1,1
top,False,False,False,False,False,False,False,False,False,False,False,False,False,False
freq,892,892,892,892,892,892,892,892,892,892,892,892,892,892


Confirming no missing values remain after flagging.

### 3.5: Custom approach

Combine the knowledge you've gained by comparing the three methods to create one final cleaned dataset where all missing values are handled optimally!

Combine the knowledge you have gained by comparing the three methods to create one final cleaned dataset where alle missing values are handled in the most optimal way!

In [64]:
titanic = pd.read_csv("titanic_uncleaned.csv", sep = ";")

titanic['Survived'] = titanic['Survived'].map({'No': 0, 'Yes' : 1})
titanic['Sex'] = titanic['Sex'].map({'male': 0, 'female' : 1})
titanic = pd.get_dummies(titanic, columns = ['Embarked'])
titanic['Embarked_C'] = titanic['Embarked_C'].astype('float')
titanic['Embarked_Q'] = titanic['Embarked_Q'].astype('float')
titanic['Embarked_S'] = titanic['Embarked_S'].astype('float')

def clean_fare(value):
    value = str(value)
    if len(value) > 3:
        if value[-3] == '.' and value[-2:].isdigit():
            value = value.replace('.', ',', 1)
        value = value.replace('.', '')
        value = value.replace(',', '.')
    return value

titanic['Fare'] = titanic['Fare'].apply(clean_fare).astype('float')

Reloading data and applying a mixed approach: drop 'Cabin', use ffill for some, median for 'Age', etc.

In [65]:
del titanic['Cabin']

Deleting the 'Cabin' column since it has too many missing values.

In [68]:
titanic['Survived'] = titanic['Survived'].ffill()
titanic['Pclass'] = titanic['Pclass'].ffill()
titanic['Name'] = titanic['Name'].fillna("")
titanic['Sex'] = titanic['Sex'].ffill()
titanic['Age'] = titanic['Age'].fillna(titanic[titanic['Age'].notna()].Age.median())
titanic['SibSp'] = titanic['SibSp'].fillna(titanic[titanic['SibSp'].notna()].Age.median())
titanic['Parch'] = titanic['Parch'].fillna(titanic[titanic['Parch'].notna()].Age.median())
titanic['Ticket'] = titanic['Ticket'].fillna("")
titanic['Fare'] = titanic['Fare'].fillna(titanic[titanic['Fare'].notna()].Age.median())

Applying a mix of ffill, fillna with median, and empty strings for optimal handling.

In [69]:
titanic.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,1,1,1,1,1,1,1,1,1,1,1,1
top,False,False,False,False,False,False,False,False,False,False,False,False,False
freq,892,892,892,892,892,892,892,892,892,892,892,892,892


Verifying that all missing values are now handled.

In [71]:
titanic.sample(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked_C,Embarked_Q,Embarked_S
760,761,0.0,3.0,"Garfirth, Mr. John",0,28.0,0.0,0.0,358585,145.00,0.0,0.0,1.0
802,803,1.0,1.0,"Carter, Master. William Thornton II",0,11.0,1.0,2.0,113760,120.00,0.0,0.0,1.0
115,116,0.0,3.0,"Pekoniemi, Mr. Edvard",0,21.0,0.0,0.0,STON/O 2. 3101294,7925.00,0.0,0.0,1.0
622,623,1.0,3.0,"Nakid, Mr. Sahid",0,20.0,1.0,1.0,2653,157417.00,1.0,0.0,0.0
101,102,0.0,3.0,"Petroff, Mr. Pastcho (\Pentcho\"")""",0,28.0,0.0,0.0,349215,78958.00,0.0,0.0,1.0
25,26,1.0,3.0,"Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...",1,38.0,1.0,5.0,347077,313875.00,0.0,0.0,1.0
209,210,1.0,1.0,"Blank, Mr. Henry",0,40.0,0.0,0.0,112277,31.00,1.0,0.0,0.0
588,589,0.0,3.0,"Gilinski, Mr. Eliezer",0,22.0,0.0,0.0,14973,8.05,0.0,0.0,1.0
211,212,1.0,2.0,"Cameron, Miss. Clear Annie",1,35.0,0.0,0.0,F.C.C. 13528,21.00,0.0,0.0,1.0
764,765,0.0,3.0,"Eklund, Mr. Hans Linus",0,16.0,0.0,0.0,347074,7775.00,0.0,0.0,1.0


Sampling 10 rows to preview the final cleaned dataset.